New: pass cwd from cli engine (#12389)
eslint/eslint: docs/developer-guide/nodejs-api.md

## Linter

The `Linter` object does the actual evaluation of the JavaScript code. It doesn't do any filesystem operations, it simply parses and reports on the code. In particular, the `Linter` object does not process configuration objects or files.
The `Linter` is a constructor, and you can create a new instance by passing in the options you want to use. The available options are:

> * `cwd` - Path to a directory that should be considered as the current working directory and can be retrieved by calling `context.getCwd()` from the rule. (If the `cwd` is `undefined`, it will be normalized to `process.cwd()` if the `process` exists, or `undefined` otherwise)

Reviewer: Wording suggestion:

```suggestion
* `cwd` - Path to a directory that should be considered as the current working directory. It is accessible to rules by calling `context.getCwd()`. If `cwd` is `undefined`, it will be normalized to `process.cwd()` if the global `process` object is defined (for example, in the Node.js runtime) , or `undefined` otherwise.
```

> * `cwd` - Path to a directory that should be considered as the current working directory and can be retrieved by calling `context.getCwd()` from the rule. (If the `cwd` is `undefined`, it will be normalized to `process.cwd()` if the `process` exists, or `undefined` otherwise)

Reviewer: Can we link to the `context.getCwd()` docs here?

For example:

```js
const Linter = require("eslint").Linter;
const linter1 = new Linter({ cwd: 'path/to/project' });
const linter2 = new Linter();
```

> in this example, rules run on `linter1` will get `path/to/project` when calling `context.getCwd()`.

Reviewer: ```suggestion
In this example, rules run on `linter1` will get `path/to/project` when calling `context.getCwd()`.
```

> Those run on `linter2` would get `process.cwd()` if the `process` exists (e.g. under node environment) or `undefined` otherwise (e.g. on the browser https://eslint.org/demo).

Reviewer: For consistency with the the previous sentence.
```suggestion
Those run on `linter2` will get `process.cwd()` if the global `process` object is defined or `undefined` otherwise (e.g. on the browser https://eslint.org/demo).
```

```js
const Linter = require("eslint").Linter;

> const linter = new Linter({});

Reviewer 2: Would you remove unnecessary change such as this line? Small diff is great thing.

Author: No problem!
Do you have any plan to remove the optional config object in the constructor?
Personally I don't think optional parameter is a good practice.

Reviewer 2: No, I don't have. Because `options` is an option, people can omit it. What are the advantages to disallow it?

Author: Personally I prefer to limit the type as much as possible.
If something is `A | undefined` then it need to be checked whether it's `A` or `undefined` for the entire scope.
It would be a heavy burden on mind especially in javascript because there is no type check.

Reviewer 2: For such a case, we have the default parameter syntax :)

For example, in [lib/cli-engine/file-enumerator.js#L190-L197](https://github.com/eslint/eslint/blob/master/lib/cli-engine/file-enumerator.js#L190-L197) case, users can omit options because of options, but every option isn't `undefined` in this scope explicitly.
